File: packages/aphrody-skills/src/frontmatter.ts
// SPDX-License-Identifier: Apache-2.0
//
// @aphrody/skills — minimal YAML front matter parser.
//
// Targets the subset that SKILL.md files actually use across the 6 upstream
// schemas (open-design, openclaw, aphrody, gemini, vercel, google). Avoids
// pulling in `js-yaml` so the package stays zero-dep.

export interface FrontMatter {
  readonly name?: string;
  readonly description?: string;
  readonly license?: string;
  readonly triggers?: readonly string[];
  readonly when_to_use?: string;
  readonly od?: {
    readonly mode?: string;
    readonly category?: string;
    readonly upstream?: string;
  };
  readonly metadata?: Record<string, string>;
  readonly raw: Record<string, unknown>;
}

const FENCE = /^---\s*\r?\n/;

/** Extract front matter + body from a SKILL.md source string. */
export function splitFrontMatter(src: string): { fm: string; body: string } {
  if (!FENCE.test(src)) {
    return { fm: "", body: src };
  }
  const afterOpen = src.replace(FENCE, "");
  const closeIdx = afterOpen.search(/\r?\n---\s*(\r?\n|$)/);
  if (closeIdx < 0) {
    return { fm: "", body: src };
  }
  const fm = afterOpen.slice(0, closeIdx);
  const body = afterOpen.slice(closeIdx).replace(/^\r?\n---\s*\r?\n?/, "");
  return { fm, body };
}

interface Line {
  readonly indent: number;
  readonly key: string;
  readonly value: string;
  readonly raw: string;
}

const stripQuotes = (s: string): string => {
  const trimmed = s.trim();
  if (trimmed.length >= 2) {
    const first = trimmed.charAt(0);
    const last = trimmed.charAt(trimmed.length - 1);
    if ((first === '"' && last === '"') || (first === "'" && last === "'")) {
      return trimmed.slice(1, -1);
    }
  }
  return trimmed;
};

const parseLine = (line: string): Line | null => {
  if (line.trim().length === 0 || line.trim().startsWith("#")) {
    return null;
  }
  const indent = line.length - line.replace(/^[ \t]+/, "").length;
  const stripped = line.slice(indent);
  if (stripped.startsWith("- ")) {
    return { indent, key: "-", value: stripped.slice(2).trim(), raw: line };
  }
  const colonIdx = stripped.indexOf(":");
  if (colonIdx < 0) {
    return null;
  }
  const key = stripped.slice(0, colonIdx).trim();
  const value = stripped.slice(colonIdx + 1);
  return { indent, key, value: value.trim(), raw: line };
};

/** Parse a YAML front-matter string (subset). Tolerant by design. */
export function parseFrontMatter(src: string): FrontMatter {
  const { fm } = splitFrontMatter(src);
  const raw: Record<string, unknown> = {};
  if (fm.length === 0) {
    return { raw };
  }

  const lines = fm.split(/\r?\n/);
  const parsed: Line[] = [];
  for (const line of lines) {
    const p = parseLine(line);
    if (p !== null) {
      parsed.push(p);
    }
  }

  let i = 0;
  while (i < parsed.length) {
    const line = parsed[i];
    if (line === undefined) {
      i += 1;
      continue;
    }
    if (line.indent !== 0 || line.key === "-") {
      i += 1;
      continue;
    }

    // Block scalar: `key: |` or `key: >` followed by an indented block.
    if (line.value === "|" || line.value === ">") {
      const block: string[] = [];
      let j = i + 1;
      while (j < parsed.length) {
        const next = parsed[j];
        if (next === undefined || next.indent === 0) {
          break;
        }
        block.push(next.raw.trim());
        j += 1;
      }
      raw[line.key] = block.join(line.value === "|" ? "\n" : " ");
      i = j;
      continue;
    }

    // Nested mapping or list: empty value with indented children.
    if (line.value.length === 0) {
      // Look ahead.
      let j = i + 1;
      const childLines: Line[] = [];
      while (j < parsed.length) {
        const next = parsed[j];
        if (next === undefined || next.indent === 0) {
          break;
        }
        childLines.push(next);
        j += 1;
      }
      if (childLines.length > 0 && childLines.every((c) => c.key === "-")) {
        raw[line.key] = childLines.map((c) => stripQuotes(c.value));
      } else if (childLines.length > 0) {
        const map: Record<string, string> = {};
        for (const c of childLines) {
          if (c.key !== "-" && c.value.length > 0) {
            map[c.key] = stripQuotes(c.value);
          }
        }
        raw[line.key] = map;
      } else {
        raw[line.key] = "";
      }
      i = j;
      continue;
    }

    raw[line.key] = stripQuotes(line.value);
    i += 1;
  }

  // Multi-line continuations: handle the open-design pattern where
  // `description: |` spans further lines that all start at column 2.
  // Already handled above; nothing to do.

  const odField = raw["od"];
  const od =
    odField && typeof odField === "object" && !Array.isArray(odField)
      ? (odField as Record<string, string>)
      : undefined;
  const triggersField = raw["triggers"];
  const triggers = Array.isArray(triggersField)
    ? (triggersField as string[])
    : undefined;
  const metadataField = raw["metadata"];
  const metadata =
    metadataField && typeof metadataField === "object" && !Array.isArray(metadataField)
      ? (metadataField as Record<string, string>)
      : undefined;

  const out: FrontMatter = {
    raw,
    ...(typeof raw["name"] === "string" ? { name: raw["name"] as string } : {}),
    ...(typeof raw["description"] === "string"
      ? { description: (raw["description"] as string).replace(/\s+/g, " ").trim() }
      : {}),
    ...(typeof raw["license"] === "string" ? { license: raw["license"] as string } : {}),
    ...(typeof raw["when_to_use"] === "string"
      ? { when_to_use: raw["when_to_use"] as string }
      : {}),
    ...(triggers ? { triggers } : {}),
    ...(od ? { od } : {}),
    ...(metadata ? { metadata } : {}),
  };
  return out;
}

File: crates/aphrody-skills/src/runtime/frontmatter.rs
// SPDX-License-Identifier: Apache-2.0
//
// SKILL.md YAML front-matter parser.
//
// Port of packages/aphrody-skills/src/frontmatter.ts. Uses `serde_yaml` so
// the full YAML 1.2 subset is supported (the TS version reimplemented a
// custom parser to stay zero-dep; Rust pulls serde_yaml from the workspace).

use std::collections::BTreeMap;

use serde::{Deserialize, Serialize};

/// Optional `od:` (open-design) metadata block recognised by several
/// upstream catalogs (open-design, openclaw, google-labs).
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct OdMeta {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub mode: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub category: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub upstream: Option<String>,
}

/// Normalised view of a SKILL.md frontmatter. Matches the TS `FrontMatter`
/// interface field-for-field (snake_case kept for compat with serde JSON
/// output that downstream callers consume).
#[derive(Debug, Clone, Default, Serialize, Deserialize, PartialEq, Eq)]
pub struct FrontMatter {
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub name: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub license: Option<String>,
    #[serde(default, skip_serializing_if = "Vec::is_empty")]
    pub triggers: Vec<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub when_to_use: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub od: Option<OdMeta>,
    #[serde(default, skip_serializing_if = "BTreeMap::is_empty")]
    pub metadata: BTreeMap<String, String>,
}

/// Split a SKILL.md source into its frontmatter (between two `---` fences)
/// and body. If no opening fence is present, returns `("", src)`.
#[must_use]
pub fn split_frontmatter(src: &str) -> (String, String) {
    // Accept CRLF or LF, optional trailing whitespace on the fence line.
    let trimmed_start = src.trim_start_matches('\u{feff}');
    let after_open = match strip_open_fence(trimmed_start) {
        Some(rest) => rest,
        None => return (String::new(), src.to_string()),
    };
    match find_close_fence(after_open) {
        Some((fm_end, body_start)) => (
            after_open[..fm_end].to_string(),
            after_open[body_start..].to_string(),
        ),
        None => (String::new(), src.to_string()),
    }
}

fn strip_open_fence(s: &str) -> Option<&str> {
    let rest = s.strip_prefix("---")?;
    // Allow trailing spaces, then require a newline.
    let rest = rest.trim_start_matches(|c: char| c == ' ' || c == '\t');
    if let Some(r) = rest.strip_prefix("\r\n") {
        return Some(r);
    }
    if let Some(r) = rest.strip_prefix('\n') {
        return Some(r);
    }
    None
}

/// Returns `(fm_end_offset, body_start_offset)` within `s`.
fn find_close_fence(s: &str) -> Option<(usize, usize)> {
    // Scan line-by-line for a line equal to `---` (with optional trailing ws).
    let mut offset: usize = 0;
    for line in s.split_inclusive('\n') {
        let line_no_eol = line.trim_end_matches('\n').trim_end_matches('\r');
        if line_no_eol.trim_end() == "---" {
            let fm_end = offset.saturating_sub(0);
            let body_start = offset + line.len();
            return Some((fm_end, body_start));
        }
        offset += line.len();
    }
    None
}

/// Parse the YAML frontmatter of a SKILL.md document. Tolerant: returns an
/// empty (default) [`FrontMatter`] if the document has no fence, or if the
/// YAML inside the fence is malformed.
#[must_use]
pub fn parse_frontmatter(src: &str) -> FrontMatter {
    let (fm, _body) = split_frontmatter(src);
    if fm.trim().is_empty() {
        return FrontMatter::default();
    }
    // Two-step parse: first into a free-form Value, then project into our
    // strongly-typed struct. This lets us normalise fields (e.g. collapse
    // whitespace in `description`) and tolerate unknown keys.
    let value: serde_yaml::Value = match serde_yaml::from_str(&fm) {
        Ok(v) => v,
        Err(_) => return FrontMatter::default(),
    };
    project(&value)
}

fn project(v: &serde_yaml::Value) -> FrontMatter {
    let mut out = FrontMatter::default();
    let map = match v {
        serde_yaml::Value::Mapping(m) => m,
        _ => return out,
    };
    for (k, val) in map {
        let key = match k.as_str() {
            Some(s) => s,
            None => continue,
        };
        match key {
            "name" => {
                out.name = val.as_str().map(str::to_string);
            }
            "description" => {
                if let Some(s) = val.as_str() {
                    // TS collapses runs of whitespace into single spaces and trims.
                    let collapsed = collapse_ws(s);
                    out.description = Some(collapsed);
                }
            }
            "license" => {
                out.license = val.as_str().map(str::to_string);
            }
            "when_to_use" => {
                out.when_to_use = val.as_str().map(str::to_string);
            }
            "triggers" => {
                if let serde_yaml::Value::Sequence(seq) = val {
                    out.triggers = seq
                        .iter()
                        .filter_map(|x| x.as_str().map(str::to_string))
                        .collect();
                }
            }
            "od" => {
                if let serde_yaml::Value::Mapping(m) = val {
                    let mut od = OdMeta::default();
                    for (kk, vv) in m {
                        if let (Some(ks), Some(vs)) = (kk.as_str(), vv.as_str()) {
                            match ks {
                                "mode" => od.mode = Some(vs.to_string()),
                                "category" => od.category = Some(vs.to_string()),
                                "upstream" => od.upstream = Some(vs.to_string()),
                                _ => {}
                            }
                        }
                    }
                    out.od = Some(od);
                }
            }
            "metadata" => {
                if let serde_yaml::Value::Mapping(m) = val {
                    for (kk, vv) in m {
                        if let (Some(ks), Some(vs)) = (kk.as_str(), vv.as_str()) {
                            out.metadata.insert(ks.to_string(), vs.to_string());
                        }
                    }
                }
            }
            _ => {}
        }
    }
    out
}

fn collapse_ws(s: &str) -> String {
    let mut out = String::with_capacity(s.len());
    let mut prev_ws = false;
    for c in s.chars() {
        if c.is_whitespace() {
            if !prev_ws && !out.is_empty() {
                out.push(' ');
            }
            prev_ws = true;
        } else {
            out.push(c);
            prev_ws = false;
        }
    }
    out.trim().to_string()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn split_no_fence_returns_empty_fm() {
        let (fm, body) = split_frontmatter("# title\nhello\n");
        assert!(fm.is_empty());
        assert_eq!(body, "# title\nhello\n");
    }

    #[test]
    fn split_extracts_fm_and_body() {
        let src = "---\nname: foo\n---\n# body\nhello\n";
        let (fm, body) = split_frontmatter(src);
        assert_eq!(fm, "name: foo\n");
        assert_eq!(body, "# body\nhello\n");
    }

    #[test]
    fn parse_basic_fields() {
        let src = "---\nname: my-skill\ndescription: A   skill\nlicense: Apache-2.0\n---\nbody\n";
        let fm = parse_frontmatter(src);
        assert_eq!(fm.name.as_deref(), Some("my-skill"));
        assert_eq!(fm.description.as_deref(), Some("A skill"));
        assert_eq!(fm.license.as_deref(), Some("Apache-2.0"));
    }

    #[test]
    fn parse_triggers_and_od() {
        let src = "---\nname: x\ntriggers:\n  - alpha\n  - beta\nod:\n  mode: design\n  upstream: github.com/foo\n---\nb\n";
        let fm = parse_frontmatter(src);
        assert_eq!(fm.triggers, vec!["alpha".to_string(), "beta".to_string()]);
        let od = fm.od.expect("od");
        assert_eq!(od.mode.as_deref(), Some("design"));
        assert_eq!(od.upstream.as_deref(), Some("github.com/foo"));
    }

    #[test]
    fn parse_tolerates_missing_fence() {
        let fm = parse_frontmatter("no fence here\n");
        assert!(fm.name.is_none());
        assert!(fm.description.is_none());
    }

    #[test]
    fn parse_handles_crlf() {
        let src = "---\r\nname: crlf\r\ndescription: line\r\n---\r\nbody\r\n";
        let fm = parse_frontmatter(src);
        assert_eq!(fm.name.as_deref(), Some("crlf"));
        assert_eq!(fm.description.as_deref(), Some("line"));
    }
}
